### DISCUSSION TITLE
Line ending normalization has occurred - PRs may have merge conflicts

### DISCUSSION THREAD

**Discussion by seanbudd** [nvaccess/nvda#16761] [Announcement]
We have just normalized the line endings for the entire repository to LF: https://github.com/nvaccess/nvda/issues/12387

This may cause merge conflicts with existing PRs.

Please handle conflicts as follows:

```sh
git checkout branchWithConflicts
# Merge master up until the normalization, handle any merge conflicts from previous code
git merge 03364662e
# Merge the line endings normalization PR, take your own unnormalized changes for now
git merge 8fb8ffcaba -X ours
# Normalize changes
git add --renormalize .
git commit -m "renormalize files"
# Merge any further changes from master, handle any merge conflicts from newer code
git merge origin/master
# Confirm diffs are as expected
git diff origin/branchWithConflicts..branchWithConflicts --ignore-cr-at-eol
git diff origin/master..branchWithConflicts --ignore-cr-at-eol
git push
```

Files can be checked out locally using CRLF if needed for Windows development using [git](https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration#_core_autocrlf).

**Comment by CyrilleB79** [comment]
Hi @seanbudd 

I have merged to the commit just before the normalization to isolate future merges from this one:
`git merge 9ea256659~1`

When I do:
```
# Merge master up until the normalization, handle any merge conflicts from previous code
git merge 9ea256659
```
and then:
`git status`
I get the following file in the changes not staged for commit (I have not modified it myself):
`        modified:   projectDocs/dev/codingStandards.md`

And then when I do:
```
# Merge the line endings normalization PR, take your own unnormalized changes for now
git merge 8fb8ffcaba -X ours
```
I logically get the following message:
```
error: Your local changes to the following files would be overwritten by merge:
        projectDocs/dev/codingStandards.md
Please commit your changes or stash them before you merge.
Aborting
```

I can probably ignore the changes in `projectDocs/dev/codingStandards.md` (there are only end-of-line CR diffs), and then ensure that this file has the correct line ending. But I'd like you to provide more systematic steps that can be applied to any branch and that can be useful for anyone.

Thanks

Cheers,

Cyrille

**Reply by seanbudd** [reply]
Hi @CyrilleB79 - I've fixed up the steps. `git merge 9ea256659` should have been `git merge 9ea256659~1` (which is 8fb8ffcaba) to prevent dirty files from being created. 

To get out of this state, you will need to `git reset --hard origin/branchWithConflicts` and try the merges again.

**Reply by CyrilleB79** [reply]
@seanbudd, I am still not able to execute this steps.

I have performed this on a specific test branch to exhibit a minimal test case:

```
# Checkout a commit before the change
nvda$ git checkout e43e7c832956d688c4ec74b1cbe8a67ea8f2e5d9
# Create from here a test branch
nvda$ git checkout -b testBranch
```

Then, I perform manually a small change, e.g. just change the copyright year in source/speech/speech.py

```
# Then add and commit your changes
nvda$ git commit -am "Change copyright year"
```

From here, I have a `testBranch` branch, with one commit diverging from master and this commit only contains a one-line change in speech/speech.py.

Then I execute the steps you have indicated:

```
# Merge master up until the normalization, handle any merge conflicts from previous code
nvda$ git merge 03364662e
# OK

# Merge the line endings normalization PR, take your own unnormalized changes for now
git merge 8fb8ffcaba -X ours
# OK

# Normalize changes
git add --renormalize .
git commit -m "renormalize files"
# OK

# Merge any further changes from master, handle any merge conflicts from newer code
git merge nvaccess/master
# NOK
```

At this step, the merge is not finished and I get a lot of files conflicted . If there should be a conflict, I would have expected it only in the file at the copyright line that I have modified.

What am I doing wrong.

Cc @mltony having had issues following these steps
Cc @nvdaes since I think that you have merged successfully some branch performing renormalization.

**Reply by nvdaes** [reply]
About comment to me (nvdaes), yes. See PR #16728

**Reply by CyrilleB79** [reply]
@nvdaes thank you for replying.

I wonder, have you faced some specific issues not described here? E.g. having more files than needed to merge.

I can see in #16728 that you have merged commit 9ea256659 on June 27, 2024. Thus it seems that you have used the steps provided by @seanbudd  before @seanbudd has edited the steps as described in https://github.com/nvaccess/nvda/discussions/16761#discussioncomment-9899242. Is it correct?

If yes, @seanbudd it's not clear to me what should be done, since I have had issues both with the steps befor you modified them and after...

For now, for the branch of PR #16825, I have just manually reported the changes in a new branch. But having a clean merge process would be better since I have other pending branches / PRs, and I am not the only one.
Thanks;

**Reply by nvdaes** [reply]
@CyrilleB79 , I followed the steps provided by Sean before he edited them, i.e., the steps of the first comment by Sean. Then I merged origin/master as approrpriate. You may want to see the log of diffs made in PR #16728

**Reply by mltony** [reply]
@CyrilleB79, FWIW, I managed to merge my PR in WSL. For more info see my comment in #16591.

**Reply by CyrilleB79** [reply]
@nvdaes what is your CrLf config?
`git config core.autocrlf`

What does it return?

**Reply by nvdaes** [reply]
@CyrilleB79 , it returns true. Ichanged this parameter recently, after line ending normalization in NVDA when working in #16728

**Comment by XLTechie** [comment]
@CyrilleB79, I successfully used a slight variant of @seanbudd's revised steps, to update PR #16369.

My variant, was that I tagged the commits he referenced, as "prenormal" and "normalize", to make doing this again easier.

"git config core.autocrlf" returns nothing.

The only problem I seem to have right now, is that I have an untracked file showing up in status, of "include/microsoft-ui-uiautomation/".
That seems harmless for this PR, so I haven't fixed it.
